router.py

```diff
@@ -44,8 +55,43 @@ def receive_packet(self, packet):
             packet (Packet): The packet we received.
 
         """
-        pass
 
-    def make_routing_table(self):
-        """Makes the routing table based on the links."""
-        pass
\ No newline at end of file
+        # If we get a routing packet, we need to update the routing table
+        if isinstance(packet, RoutingPacket):
+            self.update_routing_table(packet, link_id)
+        else: # We need to send the packet to its dest host
+            self.send_packet(packet)
+
+    def send_routing_packets(self):
+        """Sends out the routing table to all of the router's links."""
+
+        for _, link in self.links:
+            src = self.node_id
+            dest = link.get_other_node_id(self.node_id)
+            packet = RoutingPacket(None, src, dest, None, self.routing_table)
+            event = lambda: link.add_packet(packet, self.node_id)
+            self.ns.add_event(event)
+
+    def update_routing_table(self, routing_packet, link_id):
+        """Updates the routing table based on the routing packet the router just
+        received.
+
+        Args:
+            routing_packet (Packet): The routing packet we received.
+
+        """
+
+        # Check if we should update the routing table
+        changed = False
+        cost = self.links[link_id].prop_delay
+
+        for key, val in routing_packet.routing_table:
```

Comment:
1. should use routing_table.items()
2. val is a tuple, not a number
3. should rename key and val to more descriptive variable names